agent/grpc/public/services/serverdiscovery/watch_servers.go

```diff
@@ -0,0 +1,157 @@
+package serverdiscovery
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/autopilotevents"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/grpc/public"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/proto-public/pbserverdiscovery"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// We tag logs with a unique identifier to ease debugging. In the future this
+// should probably be an Open Telemetry trace ID.
+func streamID() string {
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return ""
+	}
+	return id
+}
+
+// WatchServers provides a stream on which you can receive the list of servers
+// that are ready to receive incoming requests including stale queries. The
+// current set of ready servers are sent immediately at the start of the
+// stream and new updates will be sent whenver the set of ready servers changes.
+func (s *Server) WatchServers(req *pbserverdiscovery.WatchServersRequest, serverStream pbserverdiscovery.ServerDiscoveryService_WatchServersServer) error {
+	logger := s.Logger.Named("watch-servers").With("trace_id", streamID())
+
+	logger.Debug("starting stream")
+	defer logger.Trace("stream closed")
+
+	token := public.TokenFromContext(serverStream.Context())
+
+	// Serve the ready servers from an EventPublisher subscription. If the subscription is
+	// closed due to an ACL change, we'll attempt to re-authorize and resume it to
+	// prevent unnecessarily terminating the stream.
+	var idx uint64
+	for {
```

Comment:
Not for now (as I think it'd be premature) but if we ended up writing this event-consumer-loop for a third time then it'd probably be worth extracting 😄

Reply:
👍